### DISCUSSION TITLE
Sharing thoughts on Tauri+Remix+(Deno)

### DISCUSSION THREAD

**Discussion by eric-burel** [astrodon/astrodon#51] [General]
Hey, I've just discovered Astrodon via Hacker News. It's funny because 2 weeks ago we were discussing the possibility of using Remix in Tauri and this lead me to think about the pros and cons of JS runtimes within a Tauri app.

I just wanted to share my personal notes cause they can be useful. It's specific to Remix framework, but take it as an example of mixing Rust and JS efficiently.

## Why using Tauri Remix?

- You already have a Remix app and want to make a desktop version with minimal adaptations
- You only want a desktop version now but you plan to create a SaaS/server version of your app in the future
- You struggle to find Rust developers and fullstack JavaScript developers are cheaper
- You need to produce a first version very quickly, and you prefer to learn Rust by progressively moving your backend JS code to backend Rust code.

**TL;DR: it should reduce development costs when there is no skilled Rust dev in the team**

## Why you should not use Tauri Remix?

- In a barebone Tauri app, your frontend is a Single Page Application and your backend is Tauri. Adding Remix means embedding an additional JavaScript runtime (Deno), which means a slightly bigger bundle but also that you kinda bypass Tauri.
- You don't need server-side rendering optimization in desktop software, classical SPA are perfectly fine (bundle size is not an issue and your app is usually consumed by only one person)
- It seems possible to run your Tauri app in server mode if you really need a SaaS version (to be confirmed, but in development mode you can indeed open your frontend in the browser)

**TL;DR: app will be bigger, maybe less secure, less "idiomatic" of Tauri. You probably don't need Remix.**

## Why Remix and not Next

- You don't need a separate API in a desktop app. For backend features, Remix page-level loaders and actions are more appropriate here than Next API routes.
- Next is serverless first, you would need to either need to use Next in SPA mode (next export + optionally using React Router in _app if you have dynamic routes), or to run a custom server.

## Why Deno and not Node

- Deno is written in Rust which could, potentially, enable more interesting use cases

- https://deno.com/blog/roll-your-own-javascript-runtime

### tRPC for typed communication, RSPC

See https://discord.com/channels/867764511159091230/1002486133265027152

Quoting oscartbeaumont:

"I am working on a project called rspc (https://github.com/oscartbeaumont/rspc) which is a tRPC-like server for Rust. We are planning to use it on Spacedrive (https://spacedrive.com/). However it's a very new project and probably not ready for prime time yet. I assume Tauri's Deno backend (which I recall being on the roadmap) with tRPC would be a pretty solid combination once that's ready."

"Figure I'd mention we're also working on integrating Prisma Client Rust with rspc, should make for a sweet Rust web dev experience when it's ready. https://twitter.com/brendonovichdev/status/1550981254904160257"

**Comment by marc2332** [comment]
@JonasKruckenberg might want to _attempt_ this haha
